lib/addons/karpenter/index.ts

```diff
@@ -0,0 +1,118 @@
+import { setPath } from '../../utils/object-utils'
+
+import { KarpenterControllerPolicy } from './iam'
+
+export type KarpenterAddOnProps = HelmAddOnUserProps
+
+const KARPENTER = 'karpenter'
+
+const defaultProps: HelmAddOnProps = {
+    name: KARPENTER,
+    namespace: KARPENTER,
+    version: '0.5.3',
+    chart: KARPENTER,
+    release: "ssp-addon-karpenter",
+    repository: 'https://charts.karpenter.sh',
+}
+
+export class KarpenterAddOn extends HelmAddOn {
+
+    readonly options: KarpenterAddOnProps;
+
+    // Debug Logging for Karpenter
+    readonly debugLogging: boolean;
+
+    constructor(props?: KarpenterAddOnProps) {
+        super({...defaultProps, ...props});
+        this.options = this.props;
+    }
+
+    deploy(clusterInfo: ClusterInfo): void {
+        const endpoint = clusterInfo.cluster.clusterEndpoint
+        const name = clusterInfo.cluster.clusterName
+        const cluster = clusterInfo.cluster
+        const values = { ...this.props.values ?? {} }
+
+        // Set up Node Role
+        const karpenterNodeRole = new Role(cluster, 'karpenter-node-role', {
+            assumedBy: new ServicePrincipal(`ec2.${cluster.stack.urlSuffix}`),
+            managedPolicies: [
+                ManagedPolicy.fromAwsManagedPolicyName("AmazonEKSWorkerNodePolicy"),
+                ManagedPolicy.fromAwsManagedPolicyName("AmazonEKS_CNI_Policy"),
+                ManagedPolicy.fromAwsManagedPolicyName("AmazonEC2ContainerRegistryReadOnly"),
+                ManagedPolicy.fromAwsManagedPolicyName("AmazonSSMManagedInstanceCore"),
+            ],
+            roleName: `KarpenterNodeRole-${name}`
+        });
+
+        // Set up Instance Profile
+        const karpenterInstanceProfile = new CfnInstanceProfile(cluster, 'karpenter-instance-profile', {
+            roles: [karpenterNodeRole.roleName],
+            instanceProfileName: `KarpenterNodeInstanceProfile-${name}`,
+            path: '/'
+        });
+
+        // Map Node Role to aws-auth
+        cluster.awsAuth.addRoleMapping(karpenterNodeRole, {
+            groups: ['system:bootstrapper', 'system:nodes'],
+            username: 'system:node:{{EC2PrivateDNSName}}'
+        })
+
+        // Create Namespace & SA
+        const ns = this.createNamespace(clusterInfo)
+        const sa = this.createServiceAccount(clusterInfo)
+        sa.node.addDependency(ns)
+
+        // Add helm chart
+        setPath(values, "serviceAccount.create", false)
+        setPath(values, "controller.clusterEndpoint", endpoint)
+        setPath(values, "controller.clusterName", name)
+        const karpenterChart = this.addHelmChart(clusterInfo, values, true)
+
+        karpenterChart.node.addDependency(sa);
+    }
+
+    /**
+     * Creates namespace, which is a prerequisite for service account creation and subsequent chart execution.
+     * @param clusterInfo 
+     * @returns 
+    */
+     protected createNamespace(clusterInfo: ClusterInfo): KubernetesManifest {
```

Comment:
Lets' use `utils/namespace-utils.ts#createNamespace`. It needs a slight change to support `overwrite:true` and prune. I suggest setting them to the defaults of true for all cases would be reasonable.

Reply:
+1

Reply:
Addressed.